### PR TITLE
fix: properly seed DataLoader workers for reproducibility

### LINKED ISSUE #265: Complete deterministic training by adding CUBLAS workspace config and DataLoader worker seeding
The Trainer already sets deterministic=True ([train_model.py#L319](https://github.com/mllam/neural-lam/blob/main/neural_lam/train_model.py#L319)), but two things are still missing for runs to be truly reproducible with the same seed.

1. CUBLAS_WORKSPACE_CONFIG is not set Without this env var, cuBLAS matrix multiplications on GPU can still be non-deterministic even with deterministic=True. PyTorch [documents this explicitly](https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-benchmarking). Fix is one line at the top of main():

``` python
os.environ.setdefault("CUBLAS_WORKSPACE_CONFIG", ":4096:8")
``` 
2. No worker_init_fn on the DataLoaders With num_workers=16, each worker gets a copy of the random state. Without a worker_init_fn, workers across runs can end up with correlated seeds. PyTorch [recommends](https://pytorch.org/docs/stable/data.html#randomness-in-multi-process-data-loading) seeding each worker independently:

```python
def _worker_init_fn(worker_id):
    worker_seed = torch.initial_seed() % 2**32
    np.random.seed(worker_seed)
    random.seed(worker_seed)
```
Then pass worker_init_fn=_worker_init_fn to all three DataLoaders in WeatherDataModule.

### PR DESCRIPTION
**Pull request by GiGiKoneti**
## Describe your changes

This PR addresses the `DataLoader` worker seeding part of #265 (and #267). As suggested by the maintainers in #267, PyTorch Lightning natively handles proper DataLoaders worker seeding (`worker_init_fn`) automatically if we simply pass `workers=True` to `seed_everything`. 

- Modified `neural_lam/train_model.py` to use `seed.seed_everything(args.seed, workers=True)`. This resolves the issue where multiple workers could end up with correlated random states.

*(Note: The `CUBLAS_WORKSPACE_CONFIG` environment variable from the original issue is omitted from this PR, adhering to the maintainer's decision in #267 that imposing environment variables for maximum determinism falls outside the scope of neural-lam and is better left to user execution scripts.)*

## Issue Link

Resolves #265

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [ ] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Review by sadamov — APPROVED** [review]
Ah yes, didn't know this was required, but good fix. thanks @GiGiKoneti, merging!
